### PR TITLE
Programmable HTTP mock server

### PR DESCRIPTION
**Pull request by takezoe**

### CODE REVIEW

File: airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecordStore.scala [hunk 1]
```diff
@@ -93,21 +93,21 @@ class HttpRecordStore(val recorderConfig: HttpRecorderConfig, dropSession: Boole
...
+        val headerHash = headers.map { x =>
           s"${x._1}:${x._2}".hashCode
-        }
-        .reduce { (xor, next) =>
+        }.reduce { (xor, next) =>
           xor ^ next // Take XOR to compute order-insensitive hash values.
         }
-    prefix.hashCode * 13 + headerHash
+        prefix.hashCode * 13 + headerHash
+    }
```
Comment (takezoe) [inline airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecordStore.scala]: Computing hash value hasn't worked if HeaderMap is empty.

File: airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecorder.scala [hunk 2]
```diff
@@ -99,6 +97,23 @@ object HttpRecorder extends LogSupport {
...
+    * If no matching record is found, use the given fallback handler.
+    */
+  def createProgrammableServer(programmer: HttpRecordStore => Unit): FinagleServer = {
+    val recorderConfig = HttpRecorderConfig("localhost")
+    val finagleConfig = FinagleServerConfig(recorderConfig.serverPort)
+    val recorder      = new HttpRecordStore(recorderConfig, true){
+      override def requestHash(request: Request): Int = {
+        val content = request.getContentString()
+        s"${request.method.toString()}:${recorderConfig.destAddress.hostAndPort}${request.uri}:${content.hashCode}".hashCode
+      }
```
Comment (takezoe) [inline airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecorder.scala]: It's difficult to program exact same headers. I think matching by method, address, uri and content is good enough in the programmable HTTP mock server.

File: airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecordStore.scala (hunk 1 above)
Comment (xerial) [reply airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecordStore.scala]: Nice catch!

File: airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecorder.scala (hunk 2 above)
Comment (xerial) [reply airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecorder.scala]: ok. Let's add a customization entry point to HttpRecordStore later to add user-defined pattern matching rules.

No change is required for now.

Comment (xerial) [reply airframe-http-recorder/src/main/scala/wvlet/airframe/http/recorder/HttpRecorder.scala]: My idea is adding a method like:
```
HttpRecordStore.withRequestMatcher(matcher:PartialFunction[Request, Response])
```

**Pull request merged**